apps/examples/board_specific_demo/ameba_secure_storage_demo/secure_storage.c

```diff
@@ -63,88 +69,150 @@ int sstorage_main(int argc, char *argv[])
 	if (res != SECURITY_OK) {
-		printf("security_init() Fail\n", __LINE__);
+		printf("security_init() Fail\n, %d", __LINE__);
 		sleep(1);
 	}
-
+	printf("Security init complete\n");
 	sleep(1);
-	data.data = (void *)a;
-	data.length = length;
 
-	printf("Start Secure Storage write!!\n");
-	for (slot = 0; slot < TEST_SLOT_NUM; slot++) {
-		snprintf(ss_path, 7, "ss/%d", slot);
-		memset(a, (0xA0 + slot), length);
-		a[length] = '\0';
-		printf("ss_write_secure_storage() Slot: %d\n", slot);
-		sleep(1);
-		res = ss_write_secure_storage(hnd, ss_path, 0, &data);
-		if (res != SECURITY_OK) {
-			printf("ss_write_secure_storage() Fail\n", __LINE__);
+	data.length = TEST_DATA_LENGTH;
+
+	/* Form the SS path name based on the slot index where we want to read and write */
+	snprintf(ss_path, 7, "ss/%d", slot_index);
+
+	printf("Let us read once before writing to the slot\n");
+	printf("ss_read_secure_storage() Slot: %d\n\n", slot_index);
+	res = ss_read_secure_storage(hnd, ss_path, 0, &data);
+	if (res != SECURITY_OK) {
+		printf("ss_read_secure_storage() Fail, %d, res = %d\n", __LINE__, res);
+	} else {
+		printf("data.length: %d\n", data.length);
+		memcpy(pre_read, data.data, data.length);
+		printf("data.data:\n");
+		for (i = 0; i < data.length; i++) {
+			printf("%02x ", pre_read[i]);
+			if (!((i + 1) % 16)) {
+				printf("\n");
+			}
 		}
 	}
 
-	printf("Start Secure Storage read!!\n");
 	sleep(1);
-	for (slot = 0; slot < TEST_SLOT_NUM; slot++) {
-		snprintf(ss_path, 7, "ss/%d", slot);
-		printf("ss_read_secure_storage() Slot: %d\n", slot);
-		sleep(1);
-		res = ss_read_secure_storage(hnd, ss_path, 0, &data);
-		if (res != SECURITY_OK) {
-			printf("ss_read_secure_storage() Fail\n", __LINE__);
-		} else {
-			printf("data.length: %d\n", data.length);
-			printf("data.data:\n");
-			memcpy(b, data.data, data.length);
-			for (int i = 0; i < data.length; i++) {
-				printf("%02x ", b[i]);
-				if (!((i + 1) % 16)) {
-					printf("\n");
-				}
+	printf("\nStart Flash Raw data read!!\n");
+	/* The Secure Storage Slot Flash Address , 156 = HEADER 4 + TEST_DATA_LENGTH 4 + DATA LENGTH + TAG 16 + FOOTER 4 , 1 = en_display*/
+	length = HEAD_LENGTH + DATA_LENGTH + TEST_DATA_LENGTH + TAG_LENGTH + HEAD_LENGTH;
+	if (slot_index == 0) {
+		address = SLOT0_START_ADDR;
+	} else {
+		address = SLOT1_START_ADDR + (SLOT_SIZE * (slot_index - 1));
+	}
+
+	printf("Read Raw data from Slot: %d, Area A\n", slot_index);
+	if (!rtl_ss_flash_read(address, length, pre_read, 1)) {
+		printf("rtl_ss_flash_read() Fail, %d\n", __LINE__);
+	}
+
+	sleep(1);
+	address += SLOT_AREA_OFFEST;
+	printf("\nRead Raw data from Slot: %d, Area B\n", slot_index);
+	if (!rtl_ss_flash_read(address, length, pre_read, 1)) {
+		printf("rtl_ss_flash_read() Fail, %d\n", __LINE__);
+	}
+
+	printf("Read complete\n");
+	sleep(1);
+
+	printf("Start Secure Storage write!!\n");
+	/* Populate the buffer with the user-given data */
+	memset(a, input_data, TEST_DATA_LENGTH);
+	data.data = (void *)a;
```

Comment:
The data.data is allocated at previous `ss_read_secure_storage`. But here changes the address without free of previous allocation. Could you check whether this causes memory leak or not?

Reply:
Even this is a Realtek design, we should fix the issue. Could you fix it?